### DISCUSSION TITLE
Problem with external storage and descriptors

### DISCUSSION THREAD

**Discussion by djbodya** [pimcore/pimcore#16695] [Q&A / Support]
We are using mounted network disk to store assets in out pimcore. We are using standart league/flysystem.
When I try to iterate over big amount of assets about 100k in one script it fails.

`foreach ($assetIds as $id) {
            $asset = Asset::getById($id);

            $imgData = $asset->getData();

            /*doing something*/

            $asset->save();
}
            `

Script fails after some time on $asset->getData(). It happens because of not closed descriptors.
When I try to get data using fopen fread fclose - everyrhing woks fine.

Can you help me? How i must to clear data to close descriptors?

**Comment by djbodya** [comment]
I understand what was my mistake.
Pimcore open straem te get data from the file, but pimcore close the stream only onSave, onDelete, onDestruct and several others.

Thats why if I retrive the raw data from file in a cycle and will not save, delete or destruct variable it will close all of this descriptors only when garbage colletor will work.

Why pimcore does not close file right after reading the data?

**Reply by dpfaffenbauer** [reply]
bug maybe

**Comment by nephews** [comment]
I ran into the same issue with several scripts that need to update a large amount of files (Assets).

**First solution:** run `ulimit -n 32768` before running the command. For scripts/commands that update a really large amount of files that won't work either.

**Final solution:**  that worked for me was adding `$this->closeStream();` right before the `return $this;` in `/vendor/pimcore/pimcore/models/Asset.php:610` (on Pimcore v11.5.0). This closes the open resource/file after saving.

Not sure if it messes with anything else, happy to create a PR if desired.
